nomad/periodic.go

```diff
@@ -0,0 +1,529 @@
+	if disabled {
+		return nil
+	}
+
+	// Add or update the job.
+	p.tracked[job.ID] = job
+	next := job.Periodic.Next(time.Now())
+	if tracked {
+		if err := p.heap.Update(job, next); err != nil {
+			return fmt.Errorf("failed to update job %v launch time: %v", job.ID, err)
+		}
+		p.logger.Printf("[DEBUG] nomad.periodic: updated periodic job %q", job.ID)
+	} else {
+		if err := p.heap.Push(job, next); err != nil {
+			return fmt.Errorf("failed to add job %v", job.ID, err)
+		}
+		p.logger.Printf("[DEBUG] nomad.periodic: registered periodic job %q", job.ID)
+	}
+
+	// Signal an update.
+	if p.running {
+		select {
+		case p.updateCh <- struct{}{}:
+		default:
+		}
+	}
+
+	return nil
+}
+
+// Remove stops tracking the passed job. If the job is not tracked, it is a
+// no-op.
+func (p *PeriodicDispatch) Remove(jobID string) error {
+	p.l.Lock()
+	defer p.l.Unlock()
+	return p.removeLocked(jobID)
+}
+
+// Remove stops tracking the passed job. If the job is not tracked, it is a
+// no-op. It assumes this is called while a lock is held.
+func (p *PeriodicDispatch) removeLocked(jobID string) error {
+	// Do nothing if not enabled
+	if !p.enabled {
+		return nil
+	}
+
+	if job, tracked := p.tracked[jobID]; tracked {
+		delete(p.tracked, jobID)
+		if err := p.heap.Remove(job); err != nil {
+			return fmt.Errorf("failed to remove tracked job %v: %v", jobID, err)
+		}
+	}
+
+	// Signal an update.
+	if p.running {
+		select {
+		case p.updateCh <- struct{}{}:
+		default:
+		}
+	}
+
+	p.logger.Printf("[DEBUG] nomad.periodic: deregistered periodic job %q", jobID)
+	return nil
+}
+
+// ForceRun causes the periodic job to be evaluated immediately.
+func (p *PeriodicDispatch) ForceRun(jobID string) error {
+	p.l.Lock()
+	defer p.l.Unlock()
+
+	// Do nothing if not enabled
+	if !p.enabled {
+		return fmt.Errorf("periodic dispatch disabled")
+	}
+
+	job, tracked := p.tracked[jobID]
+	if !tracked {
+		return fmt.Errorf("can't force run non-tracked job %v", jobID)
+	}
+
+	return p.createEval(job, time.Now())
+}
+
+// run is a long-lived function that waits til a job's periodic spec is met and
+// then creates an evaluation to run the job.
+func (p *PeriodicDispatch) run() {
+	defer close(p.waitCh)
+
+	// Do nothing if not enabled.
+	p.l.RLock()
+	if !p.enabled {
+		p.l.RUnlock()
+		return
+	}
+	p.l.RUnlock()
+
+	var now time.Time
+
+PICK:
+	// If there is nothing wait for an update.
+	p.l.RLock()
+	if p.heap.Length() == 0 {
+		p.l.RUnlock()
+		p.logger.Printf("[DEBUG] nomad.periodic: no periodic jobs; waiting")
+		select {
+		case <-p.stopCh:
+			return
+		case <-p.updateCh:
+		}
+		p.l.RLock()
+	}
+
+	nextJob, err := p.heap.Peek()
+	p.l.RUnlock()
+	if err != nil {
+		select {
+		case <-p.stopCh:
+			return
+		default:
+			p.logger.Printf("[ERR] nomad.periodic: failed to determine next periodic job: %v", err)
+			return
+		}
+	}
+
+	launchTime := nextJob.next
+
+	// If there are only invalid times, wait for an update.
+	if launchTime.IsZero() {
```

Comment:
Explained offline but will put it here in case anyone else is interested. You can have valid specs that will have no subsequent launches. Such as a cron spec that says launch everyday in January of 2015.